271 improve documentation for container credentials

llewellyn-sl opened the pull request:
Added provider-specific instructions for registry credential creation to:
-AWS ECR
-Azure CR
-Docker Hub
-Google AR
-Quay

Added key and access token caveats to the container registry credentials overview page. 

Added Wave context (why and where container reg credentials are to be used) to all the above.

gwright99 commented on docs/credentials/aws_registry_credentials.md:
```diff
@@ -6,18 +6,22 @@ description: 'Step-by-step instructions to set up AWS ECR credentials in Nextflo
 ![](_images/container_registry_credentials_blank.png)
 
-**4.** Enter the details of the AWS user — with the appropriate IAM role to access the registry — using the **User name** and **Password** fields.
+**4.** Enter the user's details to the **User name** and **Password** fields.
```
Is:
- User name == AWS User Access key 
- Password == AWS User Secret key

If yes, I think we need to rework this section to make it clear that:
1. Long-lived user keys must be created for this purpose.
2. The multiple ways ECR allows you to authenticate must link back to the User whose keys have been provided in Step 1.

llewellyn-sl replied:
I agree that the AWS ECR auth docs add more confusion than clarity. Removed.

Mapped user name and password fields to AWS access key fields.

Added container registry-specific credential requirement (per your point 1 above).

gwright99 commented on docs/credentials/aws_registry_credentials.md:
```diff
@@ -6,18 +6,22 @@ description: 'Step-by-step instructions to set up AWS ECR credentials in Nextflo
\ No newline at end of file
+**7.** Select **Add**. The new credential is now listed under the **Credentials** tab.
+
+**8.** In order for your pipeline execution to leverage Wave containers, add `wave { enabled=true }` to the **Nextflow config** field on the launch page. 
```
Do we need to include the other option which is to add this configuration block into the _nextflow.config_ repo file?

gwright99 commented on docs/credentials/azure_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Azure container registry crede
 ![](_images/container_registry_credentials_blank.png)
 
-**4.** Enter the credentials — to be used to access the container registry — starting with the **User name**.
+**4.** Enter `00000000-0000-0000-0000-000000000000` in the **User name** field. 
```
Is this placeholder text or do we actually expect all these 0s?

gwright99 commented on docs/credentials/azure_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Azure container registry crede
+**4.** Enter `00000000-0000-0000-0000-000000000000` in the **User name** field. 
 
-**5.** Enter the Access Key in the **Password** field.
+**5.** Enter the access token (`accessToken` — received when the `az acr login` command was run) in the **Password** field. 
```
I got an access token from Azure via `az acr login -n seqera.azurecr.io --expose-token` (using an already authenticated AZ client).

I unpacked the JWT and see that the token is only good for ~195 minutes. Does Tower / Wave refresh the token behind the scenes (meaning I only need to enter this value once at stand-up) or are there Azure expiry windows which must be considered?

pditommaso replied:
Not sure the command provided is the same ad the access keys below. These have no expiration 
<img width="1087" alt="Screenshot 2022-11-11 at 15 23 17" src="https://user-images.githubusercontent.com/816968/201360019-e3ba077a-be0e-41d9-8abb-89b849bd3009.png">

gwright99 replied:
Ah ok. Text says "Access token" so I thought I had to paste in the full JWT returned by the `az acr login` command. If long-lived "Access **keys**" are expected, that's different.

llewellyn-sl replied:
Fixed Azure steps to use long-lived credentials instead of `az acr login`-derived tokens.

gwright99 commented on docs/credentials/azure_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Azure container registry crede
-**5.** Enter the Access Key in the **Password** field.
+**5.** Enter the access token (`accessToken` — received when the `az acr login` command was run) in the **Password** field. 
 
 **6.** Enter the container registry hostname in the **Registry server** field.
```
The token already has the registry defined (e.g `"aud": "seqera.azurecr.io"`. Are we expecting this same value to be mirrored exactly in the **Registry server** field, or do we expect this root plus (optional) qualifying suffix?

gwright99 commented on docs/credentials/azure_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Azure container registry crede
\ No newline at end of file
+**7.** Select **Add**. The new credential is now listed under the **Credentials** tab.
+
+**8.** In order for your pipeline execution to leverage Wave containers, add `wave { enabled=true }` to the **Nextflow config** field on the launch page. 
```
Same comment as in AWS section. This will hold true for all entries.

llewellyn-sl replied:
My concern with removing the wave config section from each provider's subpage is that users will likely navigate directly to their provider's subpage rather than reading the container registry credentials overview page, so this information might be missed if it only lives there.

gwright99 commented on docs/credentials/docker_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Docker container registry cred
 
 ### Container registry credentials 
```
Nitpick: I suggest changing the filename to "docker_hub_registry_credentials.md" for more explicitness. When I first looked at the name I thought "is this for Docker Hub or for generic docker-type registries not otherwise called out in our docs?". 

Page content and help navigation makes it clear this is for Docker Hub, but not obvious from underlying markdown file name.

gwright99 commented on docs/credentials/docker_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Docker container registry cred
+To create your new Docker Hub registry credential in Tower, follow these steps:
 
-**2.** Enter a unique name in the **Name** field using alphanumeric characters, dashes, or underscores. 
+**1.** Create a [Personal Access Token](https://docs.docker.com/docker-hub/access-tokens/) for your user account in Docker Hub.
```
Do we need to consider the influence of 2FA? https://docs.docker.com/docker-hub/2fa/

llewellyn-sl replied:
Considered (see latest commit).

gwright99 commented on docs/credentials/docker_registry_credentials.md:
```diff
@@ -6,20 +6,26 @@ description: 'Step-by-step instructions to set up Docker container registry cred
+**6.** Enter the Personal Access Token in the **Password** field.
 
-**5.** Enter the Personal Access Token in the **Password** field.
+**7.** (Optional) Enter `docker.io` in the **Registry server** field.
```
This is optional but the AWS and Azure text did not show as optional.
Is this mismatch by accident? Or is it different due to how Docker Hub AAA works?

llewellyn-sl replied:
If the Registry server field is left empty, it defaults to docker.io. So in AWS and Azure's case, it has to be explicit.

gwright99 commented on docs/credentials/google_registry_credentials.md:
```diff
@@ -6,18 +6,28 @@ description: 'Step-by-step instructions to set up Google Cloud Artifact Registry
+From version 22.3, Tower supports the configuration of credentials for container registry services. These credentials are leveraged by the Nextflow Wave container service to authenticate to private container registries. For more information on Wave containers, see [here](https://www.nextflow.io/docs/latest/wave.html).
 
-**1.** Select **Add Credentials**. 
+Google Cloud Artifact Registry is fully integrated with Google Cloud services and integration with third party tools can also be configured, so authentication depends on your repository configuration. For further details, see [Google Artifact Registry access control](https://cloud.google.com/artifact-registry/docs/access-control). 
```
Is this entry exclusive to Google Artifact Registry? Or does it also work for Google Container Registry?

gwright99 commented on docs/credentials/quay_registry_credentials.md:
```diff
@@ -6,18 +6,22 @@ description: 'Step-by-step instructions to set up Quay container credentials in
+From version 22.3, Tower supports the configuration of credentials for container registry services. These credentials are leveraged by the Nextflow Wave container service to authenticate to private container registries. For more information on Wave containers, see [here](https://www.nextflow.io/docs/latest/wave.html).
 
-**1.** Select **Add Credentials**. 
+You grant users access to your Quay deployment using the AUTHENTICATION_TYPE specified in your [Quay configuration](https://docs.projectquay.io/config_quay.html#config-fields-required-general). Once set up, add the  credentials of a user with appropriate Quay access to Tower using these steps:
```
Docs show the following as options for AUTHENTICATION_TYPE: `Database, LDAP, JWT, Keystone, OIDC`.

Does WAVE expect a certain type of authentication or is it irrelvant by the point the user credentials are supplied?

llewellyn-sl replied:
To my understanding, the user credentials used by WAVE will authenticate to Quay as long as those credentials have the permissions/access they need. So the auth type in Quay is irrelevant to Wave, as long as that auth type grants registry access to the user credential in question.

llewellyn-sl replied:
@pditommaso is my reasoning above correct? As in, we just pass user creds to Quay and as long as those user creds have permissions to access the registry, authentication succeeds (regardless of whether the Quay configuration uses LDAP, Database etc. to determine which users have access)

pditommaso replied:
Wave expects the same authentication as it could be done via `docker login <registry host>`. 

If it works with the above command it works also for Wave.

llewellyn-sl replied:
Okay, thanks. This means that, if the user credentials can authenticate to Quay successfully, we don't care what `AUTHENTICATION_TYPE` is used in the Quay config.

pditommaso replied:
It should not. If remember correctly Quay.io has a PAT mechanism similar to Docker Hub. Think should be the proper way to configure it

llewellyn-sl replied:
They've deprecated access tokens, but they now use [Robot Accounts](https://docs.quay.io/glossary/robot-accounts.html), which is used for `docker login`. I will rewrite this section to recommend robot accounts.

llewellyn-sl replied:
Pushed update.

gwright99 commented on docs/credentials/registry_credentials.md:
```diff
@@ -6,24 +6,26 @@ description: 'Step-by-step instructions to set up container registry credentials
 
 **7.** Select **Add**. The new credential is now listed under the **Credentials** tab.
+
+**8.** In order for your pipeline execution to leverage Wave containers, add `wave { enabled=true }` to the **Nextflow config** field on the launch page. 
```
I would probably cut the generic username/password content from here - it differs for each target so probably better to just have ~"Supply credentials as described by the subpage dedicated to your specific target registry".

I'd also pull the Nextflow config part out of the subpages and centralize it here (generation of credential is different than how Tower tells Wave to use it).

gwright99 commented on docs/credentials/docker_hub_registry_credentials.md:
```diff
@@ -0,0 +1,33 @@
+From version 22.3, Tower supports the configuration of credentials for container registry services. These credentials are leveraged by the Nextflow Wave container service to authenticate to private container registries. For more information on Wave containers, see [here](https://www.nextflow.io/docs/latest/wave.html).
+
+Docker Hub makes use of Personal Access Tokens for authentication and also supports [two-factor authentication](https://docs.docker.com/docker-hub/2fa/). If you have 2FA enabled for your Docker Hub user account, you will be asked for a security code from the authenticator app configured in your account every time Tower uses your credentials to access your registry. 
+
```
I've never actually tried this. I assume someone confirmed and demonstrated this behaviour?

gwright99 reviewed (approved):
It would be nice to have 100% confirmation from Engineering re: the behaviour of Docker Hub 2FA and Quay authentication types being interchangeable from a Tower perspective. 

I already raised these questions in my previous review so I'm not going to hold up this iteration . I suggest branch not be merged until we are solid on behaviour though.

llewellyn-sl commented:
> It would be nice to have 100% confirmation from Engineering re: the behaviour of Docker Hub 2FA and Quay authentication types being interchangeable from a Tower perspective.

I already raised these questions in my previous review so I'm not going to hold up this iteration . I suggest branch not be merged until we are solid on behaviour though.

I removed the 2FA statement and added a note that we do not currently support Docker Hub credentials with 2FA enabled. Awaiting confirmation on the Quay piece before merging.
